src/box/xrow_update.c

```diff
@@ -363,8 +370,11 @@ xrow_upsert_execute(const char *expr,const char *expr_end,
 	size_t region_svp = region_used(&fiber()->gc);
 
 	if (xrow_update_read_ops(&update, expr, expr_end, format->dict,
-				 field_count) != 0)
+				 field_count) != 0) {
+		struct error *e = diag_last_error(diag_get());
+		error_set_mp(e, "ops", expr, expr_end - expr);
```

Comment:
Why are UDPATE and UPSERT handled differently? Why not set the erroneous operations in `xrow_update_read_ops` and `xrow_update_do_ops`?

Reply:
Fixed.

<details>
<summary>diff</summary>

```diff
diff --git a/src/box/xrow_update.c b/src/box/xrow_update.c
index c08a98045..92d2c85b1 100644
--- a/src/box/xrow_update.c
+++ b/src/box/xrow_update.c
@@ -106,20 +106,16 @@ xrow_update_read_ops(struct xrow_update *update, const char *expr,
 		     const char *expr_end, struct tuple_dictionary *dict,
 		     int32_t field_count_hint)
 {
-	if (mp_typeof(*expr) != MP_ARRAY) {
-		diag_set(ClientError, ER_ILLEGAL_PARAMS,
-			 "update operations must be an "
-			 "array {{op,..}, {op,..}}");
-		return -1;
-	}
+	const char *expr_begin = expr;
 	uint64_t column_mask = 0;
+	assert(mp_typeof(*expr) == MP_ARRAY);
 	/* number of operations */
 	update->op_count = mp_decode_array(&expr);
 
 	if (update->op_count > BOX_UPDATE_OP_CNT_MAX) {
 		diag_set(ClientError, ER_ILLEGAL_PARAMS,
 			 "too many operations for update");
-		return -1;
+		goto error;
 	}
 
 	if (update->op_count > 0) {
@@ -135,7 +131,7 @@ xrow_update_read_ops(struct xrow_update *update, const char *expr,
 	for (int i = 1; op < ops_end; op++, i++) {
 		if (xrow_update_op_decode(op, i, update->index_base, dict,
 					  &expr) != 0)
-			return -1;
+			goto error;
 		/*
 		 * Continue collecting the changed columns
 		 * only if there are unset bits in the mask.
@@ -223,14 +219,13 @@ xrow_update_read_ops(struct xrow_update *update, const char *expr,
 		}
 	}
 
-	/* Check the remainder length, the request must be fully read. */
-	if (expr != expr_end) {
-		diag_set(ClientError, ER_ILLEGAL_PARAMS,
-			 "can't unpack update operations");
-		return -1;
-	}
+	assert(expr == expr_end);
 	update->column_mask = column_mask;
 	return 0;
+error:
+	struct error *e = diag_last_error(diag_get());
+	error_set_mp(e, "ops", expr_begin, expr_end - expr_begin);
+	return -1;
 }
 
 /**
@@ -319,10 +314,6 @@ xrow_update_check_ops(const char *expr, const char *expr_end,
 	int ret = xrow_update_read_ops(&update, expr, expr_end,
 				       format->dict, 0);
 	region_truncate(&fiber()->gc, region_svp);
-	if (ret != 0) {
-		struct error *e = diag_last_error(diag_get());
-		error_set_mp(e, "ops", expr, expr_end - expr);
-	}
 	return ret;
 }
 
@@ -342,17 +333,18 @@ xrow_update_execute(const char *expr,const char *expr_end,
 				 field_count) != 0)
 		goto error;
 	if (xrow_update_do_ops(&update, header, old_data, old_data_end,
-			       field_count) != 0)
+			       field_count) != 0) {
+		struct error *e = diag_last_error(diag_get());
+		error_set_mp(e, "ops", expr, expr_end - expr);
+		error_set_mp(e, "tuple", header, old_data_end - header);
 		goto error;
+	}
 	if (column_mask)
 		*column_mask = update.column_mask;
 
 	return xrow_update_finish(&update, format, p_tuple_len);
 
-error:;
-	struct error *e = diag_last_error(diag_get());
-	error_set_mp(e, "ops", expr, expr_end - expr);
-	error_set_mp(e, "tuple", header, old_data_end - header);
+error:
 	region_truncate(&fiber()->gc, region_svp);
 	return NULL;
 }
@@ -370,11 +362,8 @@ xrow_upsert_execute(const char *expr,const char *expr_end,
 	size_t region_svp = region_used(&fiber()->gc);
 
 	if (xrow_update_read_ops(&update, expr, expr_end, format->dict,
-				 field_count) != 0) {
-		struct error *e = diag_last_error(diag_get());
-		error_set_mp(e, "ops", expr, expr_end - expr);
+				 field_count) != 0)
 		goto error;
-	}
 	if (xrow_upsert_do_ops(&update, header, old_data, old_data_end,
 			       field_count, suppress_error) != 0)
 		goto error;
diff --git a/test/box-luatest/error_subsystem_improvements_test.lua b/test/box-luatest/error_subsystem_improvements_test.lua
index 7964931ad..c1336e5d1 100644
--- a/test/box-luatest/error_subsystem_improvements_test.lua
+++ b/test/box-luatest/error_subsystem_improvements_test.lua
@@ -903,16 +903,35 @@ g.test_dml_update_error = function(cg)
         local test = box.schema.create_space('test')
         test:create_index('primary')
         test:replace({1, 2})
+        local ops = {}
+        for i = 1,5000 do ops[i] = {'+', 2, 1} end
+        t.assert_error_covers({
+            space = 'test',
+            space_id = test.id,
+            index = 'primary',
+            index_id = 0,
+            ops = ops,
+            message = 'Illegal parameters, too many operations for update'
+        }, test.update, test, {1}, ops)
         t.assert_error_covers({
             space = 'test',
             space_id = test.id,
             index = 'primary',
             index_id = 0,
-            tuple = {1, 2},
             ops = {{'+', 2, 'foo'}},
             message = "Argument type in operation '+' on field 2 does not " ..
                       "match field type: expected a number",
         }, test.update, test, {1}, {{'+', 2, 'foo'}})
+        t.assert_error_covers({
+            space = 'test',
+            space_id = test.id,
+            index = 'primary',
+            index_id = 0,
+            ops = {{':', 2, 2, 1, 'foo'}},
+            tuple = {1, 2},
+            message = "Argument type in operation ':' on field 2 does not " ..
+                      "match field type: expected a string"
+        }, test.update, test, {1}, {{':', 2, 2, 1, 'foo'}})
 
         local dummy = function(_, new) return new end
         test:before_replace(dummy)
@@ -921,7 +940,6 @@ g.test_dml_update_error = function(cg)
             space_id = test.id,
             index = 'primary',
             index_id = 0,
-            tuple = {1, 2},
             ops = {{'+', 2, 'foo'}},
             message = "Argument type in operation '+' on field 2 does not " ..
                       "match field type: expected a number"
@@ -936,7 +954,6 @@ g.test_dml_update_error = function(cg)
             space_id = space.id,
             index = 'primary',
             index_id = 0,
-            tuple = {'sql_full_metadata', false},
             ops = {{'+', 2, 'foo'}},
             message = "Argument type in operation '+' on field 2 does not " ..
                       "match field type: expected a number"
@@ -950,7 +967,6 @@ g.test_dml_update_error = function(cg)
             space_id = test.id,
             index = 'primary',
             index_id = 0,
-            tuple = {1, 2},
             ops = {{'+', 2, 'foo'}},
             message = "Argument type in operation '+' on field 2 does not " ..
                       "match field type: expected a number",

```